### ISSUE TITLE
Consider getting rid of or replacing `semver.js`

### ISSUE THREAD

**Issue by p-himik** [thheller/shadow-cljs#998]
Clojurians Slack discussion thread: https://clojurians.slack.com/archives/C6N245JGG/p1647112910767939

Options:
- Keep parsing and comparing ranges via one of the Java SemVer libraries
- Just install the highest version and hope for the best
- Install something according to one found version range and warn if there are other version ranges that aren't byte-for-byte the same
- Like the above but make it an interactive command

According to what I gather from `npm` docs, there seems to be yet another option.
You can concatenate versions with spaces, it won't break anything since there's only one binary operation, `||`, and it has the highest precedence - the spaces are implied as `&&`. So something like `>1 <8 >2 >3 || <1` is a valid version string.
So:
- Gather all `:npm-deps`
- `merge-with` them with concatenating the version strings with spaces (replace empty versions with `"*"`)
- Add to those the contents of `package.json`, but only for dependencies that were in `:npm-deps`
- Run `npm install` for every accumulated dependency with an explicit version
- If there are conflicts, `npm` will loudly fail. If there are no conflicts, the required dependencies will be installed, if they haven't been installed before.

@thheller has mentioned that "forking a node process has other issues". I don't know what those issues are, but `npm` does use `node`, and shadow-cljs does shell out to `npm` when installing dependencies.

**Comment by thheller** [comment]
In `2.18.0` I have removed support for the `semver` checks entirely and just use the first `deps.cljs` `:npm-deps` version specified for each package. This is the simplest solution that allows me to get rid of the graal-js engine.

I might consider reworking this again at some point but for now this should be good enough.

If anyone is seriously interested in working on this you could totally do this as a completely separate library. Nothing in this is dependent on shadow-cljs. As a separate lib/tool this could also then be used together with `figwheel` or just `cljs.main`. It is just a separate command that needs to get the `deps.cljs` files from the classpath ([as done here, feel free to copy](https://github.com/thheller/shadow-cljs/blob/6daffeae50d0b7f029aad8e975997b20223955dc/src/main/shadow/cljs/devtools/server/npm_deps.clj#L130-L148)) and then install the packages somehow. You could use the entire `npm_deps.clj` namespace as a reference.

**Comment by p-himik** [comment]
There's a slight leftover: https://github.com/thheller/shadow-cljs/blob/c2d8e13a562f3a774720fb130f172e42896eb8ca/src/main/shadow/cljs/npm/check_versions.clj#L42

No clue whether it affects anything - I just noticed it by accident.

**Comment by thheller** [comment]
Good catch. That used to be part of the `:browser` build process checking the installed version but ended up more noise than actual usable info. I removed it a while ago and made it a separate runnable `-main`. I don't think I ever told anyone about this so I'll just remove it entirely.

Build reports are also much better for understanding your builds anyways.
